### ISSUE TITLE
PgUpgrade stuck on progressing state

### ISSUE THREAD

**Issue by Martin-Hogge** [CrunchyData/postgres-operator#3518]
- Platform: Kubernetes 1.22.15-gke.100 
- PGO Image Tag: ubi8-5.1.0-0
- Postgres version: 13 -> 14

My cluster is actually running with Postgres 13 but I would like to upgrade to Postgres 14.
In order to do so I followed these instructions: https://access.crunchydata.com/documentation/postgres-operator/v5/guides/major-postgres-version-upgrade/

Everything is working so far but the `pgupgrade`  seems stuck in a `progressing` state.
Here is a `describe` of the `pgupgrade`:

```
Name:         upgrade-postgres-adisoft-dev-02
Namespace:    adisoft-dev
Labels:       <none>
Annotations:  <none>
API Version:  postgres-operator.crunchydata.com/v1beta1
Kind:         PGUpgrade
Metadata:
  Creation Timestamp:  2023-01-03T09:19:04Z
  Generation:          1
  Managed Fields:
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:fromPostgresVersion:
        f:image:
        f:postgresClusterName:
        f:toPostgresVersion:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2023-01-03T09:19:04Z
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:conditions:
          .:
          k:{"type":"Progressing"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
    Manager:         pgupgrade-controller
    Operation:       Update
    Subresource:     status
    Time:            2023-01-03T09:19:04Z
  Resource Version:  550227610
  UID:               60a57424-b8ae-4072-841f-f15d1a73e962
Spec:
  From Postgres Version:  13
  Image:                  registry.developers.crunchydata.com/crunchydata/crunchy-upgrade:ubi8-5.1.0-0
  Postgres Cluster Name:  postgres-adisoft-dev
  To Postgres Version:    14
Status:
  Conditions:
    Last Transition Time:  2023-01-03T09:21:25Z
    Message:               Upgrade progressing for cluster postgres-adisoft-dev
    Observed Generation:   1
    Reason:                PGUpgradeProgressing
    Status:                True
    Type:                  Progressing
Events:                    <none>
```

What seems strange is that the `job` has completed successfully:
```
NAME                                     COMPLETIONS   DURATION   AGE
postgres-adisoft-dev-backup-h87d         1/1           105s       90m
upgrade-postgres-adisoft-dev-02-pgdata   1/1           24s        61m
```

Logs of the job `upgrade-postgres-adisoft-dev-02-pgdata` finish with: 
```
pg_upgrade Job Complete!
```

Here is a `describe` of the `postgrescluster`:
```
Name:         postgres-adisoft-dev
Namespace:    adisoft-dev
Labels:       <none>
Annotations:  postgres-operator.crunchydata.com/allow-upgrade: upgrade-postgres-adisoft-dev-02
API Version:  postgres-operator.crunchydata.com/v1beta1
Kind:         PostgresCluster
Metadata:
  Creation Timestamp:  2023-01-03T08:52:37Z
  Finalizers:
    postgres-operator.crunchydata.com/finalizer
  Generation:  2
  Managed Fields:
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:backups:
          .:
          f:pgbackrest:
            .:
            f:configuration:
            f:global:
              .:
              f:repo1-path:
              f:repo1-retention-full:
              f:repo1-retention-full-type:
              f:repo1-s3-uri-style:
            f:image:
            f:repos:
              .:
              k:{"name":"repo1"}:
                .:
                f:name:
                f:s3:
                  .:
                  f:bucket:
                  f:endpoint:
                  f:region:
                f:schedules:
                  .:
                  f:full:
        f:image:
        f:instances:
          .:
          k:{"name":"instance1"}:
            .:
            f:affinity:
              .:
              f:podAntiAffinity:
                .:
                f:requiredDuringSchedulingIgnoredDuringExecution:
            f:dataVolumeClaimSpec:
              .:
              f:accessModes:
              f:resources:
                .:
                f:requests:
                  .:
                  f:storage:
            f:metadata:
              .:
              f:annotations:
                .:
                f:cluster-autoscaler.kubernetes.io/safe-to-evict:
              f:labels:
                .:
                f:app:
            f:name:
            f:replicas:
            f:walVolumeClaimSpec:
              .:
              f:accessModes:
              f:resources:
                .:
                f:requests:
                  .:
                  f:storage:
        f:monitoring:
          .:
          f:pgmonitor:
            .:
            f:exporter:
              .:
              f:image:
        f:patroni:
          .:
          f:dynamicConfiguration:
            .:
            f:postgresql:
          f:leaderLeaseDurationSeconds:
          f:port:
          f:syncPeriodSeconds:
        f:port:
        f:postgresVersion:
        f:users:
          .:
          k:{"name":"adisoft"}:
            .:
            f:databases:
              .:
              v:"adisoft":
            f:name:
            f:options:
          k:{"name":"bv"}:
            .:
            f:databases:
              .:
              v:"adisoft":
            f:name:
            f:options:
          k:{"name":"mh"}:
            .:
            f:databases:
              .:
              v:"adisoft":
            f:name:
            f:options:
          k:{"name":"nh"}:
            .:
            f:databases:
              .:
              v:"adisoft":
            f:name:
            f:options:
          k:{"name":"pdc"}:
            .:
            f:databases:
              .:
              v:"adisoft":
            f:name:
          k:{"name":"tg"}:
            .:
            f:databases:
              .:
              v:"adisoft":
            f:name:
            f:options:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2023-01-03T08:52:37Z
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"postgres-operator.crunchydata.com/finalizer":
    Manager:      postgrescluster-controller
    Operation:    Update
    Time:         2023-01-03T08:52:37Z
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        f:shutdown:
    Manager:      kubectl-patch
    Operation:    Update
    Time:         2023-01-03T09:19:47Z
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:conditions:
          .:
          k:{"type":"PGBackRestReplicaCreate"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
          k:{"type":"PGBackRestReplicaRepoReady"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
        f:databaseRevision:
        f:instances:
          .:
          k:{"name":"instance1"}:
            .:
            f:name:
        f:monitoring:
          .:
          f:exporterConfiguration:
        f:observedGeneration:
        f:patroni:
          .:
          f:systemIdentifier:
        f:pgbackrest:
          .:
          f:repos:
            .:
            k:{"name":"repo1"}:
              .:
              f:name:
              f:replicaCreateBackupComplete:
              f:repoOptionsHash:
              f:stanzaCreated:
        f:proxy:
          .:
          f:pgBouncer:
            .:
            f:postgresRevision:
        f:startupInstance:
        f:startupInstanceSet:
        f:usersRevision:
    Manager:      postgrescluster-controller
    Operation:    Update
    Subresource:  status
    Time:         2023-01-03T09:19:48Z
    API Version:  postgres-operator.crunchydata.com/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:postgres-operator.crunchydata.com/allow-upgrade:
    Manager:         kubectl-annotate
    Operation:       Update
    Time:            2023-01-03T09:21:25Z
  Resource Version:  550227606
  UID:               caae92f1-07cc-4d96-a2fa-2c3e41031b12
Spec:
  Backups:
    Pgbackrest:
      Configuration:
        Secret:
          Name:  pgo-s3-creds
      Global:
        repo1-path:                 /pgbackrest/adisoft-dev/postgres-adisoft-dev/repo1
        repo1-retention-full:       1
        repo1-retention-full-type:  time
        repo1-s3-uri-style:         path
      Image:                        registry.developers.crunchydata.com/crunchydata/crunchy-pgbackrest:ubi8-2.38-0
      Repos:
        Name:  repo1
        s3:
          Bucket:    adisoft-dev-postgres-backup
          Endpoint:  backupstore.adisoft.be
          Region:    us-east-1
        Schedules:
          Full:  1 1 1 1 1
  Image:         europe-west1-docker.pkg.dev/helical-parity-286714/adisoft-registry/adi-crunchy:v4
  Instances:
    Affinity:
      Pod Anti Affinity:
        Required During Scheduling Ignored During Execution:
          Label Selector:
            Match Labels:
              postgres-operator.crunchydata.com/cluster:       postgres-adisoft-dev
              postgres-operator.crunchydata.com/instance-set:  instance1
          Topology Key:                                        kubernetes.io/hostname
    Data Volume Claim Spec:
      Access Modes:
        ReadWriteOnce
      Resources:
        Requests:
          Storage:  20Gi
    Metadata:
      Annotations:
        cluster-autoscaler.kubernetes.io/safe-to-evict:  true
      Labels:
        App:   postgres-adisoft-dev
    Name:      instance1
    Replicas:  3
    Wal Volume Claim Spec:
      Access Modes:
        ReadWriteOnce
      Resources:
        Requests:
          Storage:  10Gi
  Monitoring:
    Pgmonitor:
      Exporter:
        Image:  registry.developers.crunchydata.com/crunchydata/crunchy-postgres-exporter:ubi8-5.3.0-0
  Patroni:
    Dynamic Configuration:
      Postgresql:
        Parameters:
          synchronous_commit:         on
          synchronous_standby_names:  *
        pg_hba:
          hostnossl all all all md5
    Leader Lease Duration Seconds:  30
    Port:                           8008
    Sync Period Seconds:            10
  Port:                             5432
  Postgres Version:                 13
  Shutdown:                         true
  Users:
    Databases:
      adisoft
    Name:     adisoft
    Options:  SUPERUSER
    Databases:
      adisoft
    Name:     mh
    Options:  SUPERUSER
    Databases:
      adisoft
    Name:     bv
    Options:  SUPERUSER
    Databases:
      adisoft
    Name:     nh
    Options:  SUPERUSER
    Databases:
      adisoft
    Name:     tg
    Options:  SUPERUSER
    Databases:
      adisoft
    Name:  pdc
Status:
  Conditions:
    Last Transition Time:  2023-01-03T08:53:08Z
    Message:               pgBackRest replica create repo is ready for backups
    Observed Generation:   2
    Reason:                StanzaCreated
    Status:                True
    Type:                  PGBackRestReplicaRepoReady
    Last Transition Time:  2023-01-03T08:54:54Z
    Message:               pgBackRest replica creation is now possible
    Observed Generation:   2
    Reason:                RepoBackupComplete
    Status:                True
    Type:                  PGBackRestReplicaCreate
  Database Revision:       86f54dc978
  Instances:
    Name:  instance1
  Monitoring:
    Exporter Configuration:  6cf9b7bb76
  Observed Generation:       2
  Patroni:
    System Identifier:  7184346298476863575
  Pgbackrest:
    Repos:
      Name:                            repo1
      Replica Create Backup Complete:  true
      Repo Options Hash:               75d4cfc7d
      Stanza Created:                  true
  Proxy:
    Pg Bouncer:
      Postgres Revision:  5c9966f6bc
  Startup Instance:       postgres-adisoft-dev-instance1-xlgv
  Startup Instance Set:   instance1
  Users Revision:         78c6c64859
Events:                   <none>
```

I already tried to upgrade this cluster from 13 to 14 one time before and I had the same issue -> job completed by `pgupgrade` "progressing". Last time I didn't wait for the `pgupgrade` to complete and reboot the cluster but then only one pod sucessfully restarted: the 2 others (replicas) crashed on errors like 

> The data directory was initialized by PostgreSQL version 13, which is not compatible with this version 14.2.

And 

> system ID mismatch, node postgres-adisoft-dev-instance1-zfl7-0 belongs to a different cluster: 7184336454194233377 != 7184069716819869788

This is why I'm now waiting for the `pgupgrade` to suceed but after more than 1h on a fresh cluster it doesn't seems right to me.

Any idea of what I did wrong or what I should do to sucessfully upgrade my postgres version from 13 to 14 ?

Thank you.

**Comment by benjaminjb** [comment]
To double-check, what are the logs for the `pgo-upgrade-<SOMETHING>` pod? If you're using PGO v5.1.x, the upgrade controller is its own pod, so it should be relatively easy to see the logs.

The only thing that jumps out at me right now is that your postgrescluster has`replicas: 3` -- which is fine, pgo-upgrade should be able to handle that. But with replicas, the behavior for upgrading should be to create a job to upgrade the primary db and a job to remove the data on the secondaries. So I'm not surprised that you're getting errors like

```
The data directory was initialized by PostgreSQL version 13, which is not compatible with this version 14.2.
```

It looks like the job to remove the data from the secondaries is not running. Which is why I'd want to see the logs for the upgrade controller.

**Comment by Martin-Hogge** [comment]
Ok my bad I didn't saw that their was a `pgo-upgrade` pod (it's not on the same namespace as the cluster I try to upgrade). By reading the logs I find out what was the issue: 


> 2023-01-04T05:10:21.815Z ERROR controller-runtime.manager.controller.pgupgrade Reconciler error {"reconciler group": "postgres-operator.crunchydata.com", "reconciler kind": "PGUpgrade", "name": "upgrade-postgres-adisoft-dev-02", "namespace": "adisoft-dev", "error": "Job.batch \"upgrade-postgres-adisoft-dev-02-postgres-adisoft-dev-instance1-5b4w\" is invalid: spec.template.labels: Invalid value: \"upgrade-postgres-adisoft-dev-02-postgres-adisoft-dev-instance1-5b4w\": must be no more than 63 characters"}

I recreated the cluster and retry with a shorter name (`pgupgrade-03`) and in fact it created 2 more jobs (for replicas) but those two failed with the logs below

> Removing PostgreSQL data dir for pg13...
> Checking the directory exists and isn't being used...
> Removing old pgdata directory...
> rm: cannot remove '/pgdata/pg13_wal': No such file or directory

I checked on another cluster (using pg13 too) and the wal directory is at `/pgdata/pg13/pg_wal`. Is their a way to tell `pgupgrade` where it should find the wal directory ? I check the spec but didn't find anything there.

**Comment by benjaminjb** [comment]
> Ok my bad I didn't saw that their was a `pgo-upgrade` pod (it's not on the same namespace as the cluster I try to upgrade)

Depending on the installation configuration, I would expect the `pgo-upgrade` pod to be in the same namespace as the `pgo` pod, so that's expected.

But this is unexpected:

> I checked on another cluster (using pg13 too) and the wal directory is at `/pgdata/pg13/pg_wal`. Is their a way to tell `pgupgrade` where it should find the wal directory ? I check the spec but didn't find anything there.

OK, first the hacky workaround: if you need to upgrade, you could decrease replicas so you only have a primary, and then upgrade. I think that would work -- since the primary job seemed to finish for you without problem -- but it's hacky and not the way this is supposed to work.

(But sometimes it's nice to know you can do the workaround if it becomes necessary. Although why would )

What's confusing to me is I thought `/pgdata/pg13/pg_wal` would be symlinked to `/pgdata/pg13_wal`, which is what I see when I just start up a new cluster and exec in:

```
bash-4.4$ ls -al /pgdata/pg13/pg_wal
lrwxrwxrwx 1 postgres postgres 16 Jan  4 17:26 /pgdata/pg13/pg_wal -> /pgdata/pg13_wal
```

But wait! We added a fix in v5.1.1 : can you check the rmdata job pod that failed, does the command have this line: 

`rm -r /pgdata/pg"${old_version}" /pgdata/pg"${old_version}_wal"`

or 

`rm -rf /pgdata/pg"${old_version}" "$(realpath /pgdata/pg${old_version}/pg_wal)"`?

If it has the first, then I think updating to a newer version than 5.1.0 would work -- and I'd be curious if it did. (That would let me know that this problem is this bug, which we fixed in later versions.)

**Comment by Martin-Hogge** [comment]
Upgrading to v5.1.1 fixed it !

Thank you for your support @benjaminjb
